### ISSUE TITLE
New pyproj.CRS based interface for the .crs in GeoPandas

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#1003]
There has been some discussion lately about pyproj / the projection library dependency, eg https://github.com/geopandas/geopandas/pull/885 (which proposed to remove pyproj dependency in favor of fiona, which also relates to https://github.com/geopandas/geopandas/issues/315) and https://github.com/geopandas/geopandas/issues/965 about pyproj 2 compatibility.

**~Short term~**: we need to guarantee compatibility with the new proj 6 / pyproj 2. This already has been done in https://github.com/geopandas/geopandas/pull/962. I *think* this should fully guarantee backwards compatibility in user behaviour, given that we still use proj4 strings internally to store the CRS.

**Long term**:

* We should *at least* update the *internal* representation of the CRS, to no longer use the sub-optimal proj4 strings. Similarly to what rasterio has done (and fiona will do), we could use the WKT instead as the canonical format (https://github.com/mapbox/rasterio/pull/1597).

* In addition, we could also more fully adopt the new `pyproj.CRS` (pyproj >= 2) class as the internal and *user facing* representation of the CRS: `gdf.crs` would return a `pyproj.CRS` class. 
  This would give a more complete user experience: instead of a "dumb" string, we get access to a rich CRS class that can be inspected (eg is it projected?) or converted to other formats

@snowman2 started a PR for this: https://github.com/geopandas/geopandas/pull/998. But let's use this issue for the general discussion.

cc @geopandas/collaborators

**Comment by jorisvandenbossche** [comment]
I personally think having an actual, more user friendly CRS class instead of the current proj4 string or dict is an important new feature for GeoPandas in terms of usability / user friendliness. 

To get an idea of this CRS class: http://pyproj4.github.io/pyproj/dev/api/crs.html

But some questions about the approach:

* Are we OK with relying on another library (pyproj) instead of having the CRS class definition in geopandas itself?
* If we rely on pyproj, are we OK with directly requiring pyproj > 2 / proj 6?
  * I personally think it is better to require this, so that the return type of `GeoDataFrame.crs` does not depend on the version of pyproj being installed (but rather on the geopandas version: so eg geopandas >= 0.6 always returns a CRS objects instead of dict/string)

Some feedback very welcome!

**Comment by snowman2** [comment]
I would recommend pinning `pyproj >=2.2.0` after it is released if you choose to use the `pyproj.CRS`. This will make the transition smoother and the behavior of `geopandas` consistent. Otherwise. depending on the version of `pyproj` transformations or the `crs` variable will have different results. For example, `pyproj>=2.2.0` takes into account the new axis swapping that occurs in PROJ (see: https://pyproj4.github.io/pyproj/dev/api/transformer.html#transformer) and will prevent user confusion.


> Are we OK with relying on another library (pyproj) instead of having the CRS class definition in geopandas itself?

I am biased, but I definitely would like to see `geopandas` use the `pyproj.CRS` class :smile:. If not, `fiona` is planning likely have a CRS class in the 2.0 release (https://github.com/Toblerity/Fiona/issues/714). Either way, I think it would be a good idea for `geopandas` to use one of the `CRS` classes in their dependencies.

**Comment by sgillies** [comment]
@jorisvandenbossche @snowman2 one big downside to using pyproj in geopandas is that your project will now have two C extension dependencies that link with PROJ. This exposes your users to issues of binary compatibility and will increase the cost of "customer" support. At least that is my experience from supporting shapely, fiona, and rasterio, and this is why I told @snowman2 that I was 👎 on having fiona and rasterio depend on pyproj. I'm not saying "stop", only saying that you should consider this carefully.

**Comment by snowman2** [comment]
@sgillies, pyproj and Fiona have been dependencies in geopandas for a long time now. Also, there are multiple projects that use both of them out there. I think instead of avoiding using them together due to potential binary wheel issues, we could look for ways to make them work better together.

Also, I think geopandas recommends using conda for installation, and that is a non-issue there.

**Comment by jorisvandenbossche** [comment]
As @snowman2 says, pyproj is already a dependency of geopandas for a long time (for doing the actual coordinate conversion, this issue proposes to integrate more tightly with pyproj to also use their new CRS class). So we already have this problem.

In general, I think those multiple dependencies to the same C library will get unavoidable. Unless fiona would subsume all functionality of pyproj, for this case (but which I think is not desrable). But also for GEOS we are going in that direction.

**Comment by JimShady** [comment]
Just come across this discussion. Using to_crs is incredibly slow in GeoPandas. Is there a recommended way to improve this? A specific version of pyproj perhaps, or have all improvements already been implemented?

**Comment by snowman2** [comment]
@JimShady, are you using the latest version of geopandas (I believe it is 0.5)?

**Comment by JimShady** [comment]
@snowman2 

    geopandas.__version__
    '0.4.1'

    pyproj.__version__
    '2.2.1'

I should update geopandas from Github?

**Comment by martinfleis** [comment]
> I should update geopandas from Github?
0.5.0 is on conda-forge and PIP, you don't need to install master from GH.

**Comment by snowman2** [comment]
You should be able to upgrade from pypi. The new version is much faster.

**Comment by jorisvandenbossche** [comment]
@JimShady can you open a new issue for this? (this issue is only about the user interface, not about the performance)

**Comment by snowman2** [comment]
I am realizing the problem. I think there are some metadata issues on conda-forge that caused your problem. I would re-create your environment with a pin for geopandas.

**Comment by JimShady** [comment]
> @JimShady can you open a new issue for this? (this issue is only about the user interface, not about the performance)

https://github.com/geopandas/geopandas/issues/1055

**Comment by martinfleis** [comment]
To refresh this discussion, in #1095 is a nice example how would implementation of pyproj.CRS help. 

> Are we OK with relying on another library (pyproj) instead of having the CRS class definition in geopandas itself?

It is required dependency, it is maintained, we rely on it for a long time. It is a bit risky as pyproj might change in a way geopandas might not like, but as @snowman2 is quite active here I would be OK with it. It can provide much richer functionality than geopandas itself can offer.

> If we rely on pyproj, are we OK with directly requiring pyproj > 2 / proj 6?

Yes.

In general, I support using `pyproj.CRS` class for `gdf.crs`. I always felt that dict geopandas returns is a bit unfit.

@snowman2 can you guess how much work needs to be done in #998 to make it ready if we'll give it a go?

**Comment by snowman2** [comment]
> can you guess how much work needs to be done in #998 to make it ready if we'll give it a go?

If you want the pyproj.CRS instead of the dict, just need to delete the second commit and it is ready to go. Maybe some minor updates if conflicts exist from any recent changes.

**Comment by snowman2** [comment]
If you want to require a specific version of pyproj, then some cleanup might be needed.

**Comment by jorisvandenbossche** [comment]
I think the main blocker for directly requiring pyproj > 2 / PROJ > 6, is the ecosystem adoption. 

Doing a `conda create -n test-geo geopandas cartopy rasterio` still gets pyproj 1.9.6

**Comment by martinfleis** [comment]
We are blocked by both cartopy and rasterio. @snowman2's PR #998 is conditional, so it should work in envs with rasterio (hence pyproj 1.9.6) and without (pyproj>2).

Can we go ahead with it so our development is not blocked by others? We will keep compatibility, just new features will be limited if you want to use rasterio/cartopy as well, until they will resolve their issues with PROJ4.  It will give us not only nice CRS handling, but also the ability to do distance-based stuff on unprojected CRS (like great circle) or warn if you want to use method which is supported only for projected coords on unprojected one.

**Comment by jorisvandenbossche** [comment]
My personal preference was to switch completely, when starting to use it. So meaning requiring pyproj > 2. I wanted to avoid that the output of `gdf.crs` would be dependent on the version of pyproj.

But if it takes longer for other projects to update, then maybe should already start to use optionally.

**Comment by snowman2** [comment]
Update on cartopy/rasterio:
- `cartopy` is currently supported by PROJ 6+ on the `conda-forge` channel and merged in the change to support PROJ 6 [here](https://github.com/SciTools/cartopy/pull/1289).  
- `rasterio` just released a version compatible with GDAL 3+  & PROJ 6+ (1.0.25) [release notes](https://github.com/mapbox/rasterio/blob/096cb4878fd3490bad52e17835285de8a2aeeebf/CHANGES.txt). Currently the `rasterio` build is in progress on conda-forge [here](https://github.com/conda-forge/rasterio-feedstock/pull/136).

**Comment by snowman2** [comment]
Another update for `rasterio`:
 - Build passing for GDAL 2.4 & PROJ 6+ [here](https://github.com/conda-forge/rasterio-feedstock/pull/138). So, this will should resolve the main issues going with PROJ 6+.

**Comment by snowman2** [comment]
On another note, I just noticed that PROJ 6.1.1 was installed with this distribution:

```
FROM python:3.6-stretch

ENV container docker

# Update base container install
RUN apt-get update
RUN apt-get upgrade -y

# Add unstable repo to allow us to access latest GDAL builds
RUN echo deb http://ftp.uk.debian.org/debian unstable main contrib non-free >> /etc/apt/sources.list
RUN apt-get update

# Existing binutils causes a dependency conflict, correct version will be installed when GDAL gets intalled
RUN apt-get remove -y binutils

# Install GDAL dependencies
RUN apt-get -t unstable install -y libgdal-dev g++ gdal-bin
```
Based on: https://github.com/thinkWhere/GDAL-Docker

**Comment by martinfleis** [comment]
Update on the state of geo stack.

Doing `conda create -n test-geo geopandas cartopy rasterio` using default channel will get pyproj1.9.6 (and also Geopandas 0.4.1), while `conda create -n test-geo geopandas cartopy rasterio -c conda-forge` will give pyproj 2.3.1. 

So it seems to me that we are ready to roll out CRS class as there should not be dependency clash anymore (using conda-forge).

**Comment by jorisvandenbossche** [comment]
The default channel is indeed lagging behind in general, so not sure we can wait on them (I opened https://github.com/ContinuumIO/anaconda-issues/issues/11400 about it)

So maybe we should go forward with this? 
Are we OK with doing the hard version requirement?

**Comment by martinfleis** [comment]
> So maybe we should go forward with this?

YES!

> Are we OK with doing the hard version requirement?

Yes.

**Comment by snowman2** [comment]
pyproj 2.4.1 is now available: https://anaconda.org/anaconda/pyproj
Also, installed Fiona with GDAL 3.0.2 using defaults.
